asserts/tool/sign.go

```diff
@@ -0,0 +1,221 @@
+ *
+ */
+
+// Package tool offers tooling to sign assertions.
+package tool
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/snapcore/snapd/asserts"
+)
+
+// The supported media types for the input of assertion signing.
+const (
+	JSONInput = "application/json"
+	YAMLInput = "application/x-yaml"
+)
+
+// SignRequest lets specify the complete input for signing an assertion.
+type SignRequest struct {
+	// the key to use can be speficied either passing the text of
+	// an account-key assertion
+	AccountKey []byte
+	// or passing the key id
+	KeyID string
+	// and an optional account-id of the signer (if left out headers value are consulted)
+	AuthorityID string
+
+	// the assertion type (as a string)
+	AssertionType string
+	// specify the media type of the input
+	StatementMediaType string
+	// Statement is used as input to construct the assertion
+	// it's a mapping encoded as either JSON or YAML (specified in StatementMediaType)
+	// either of just the header fields of the assertion
+	// or containting exactly two entries
+	// "headers": mapping with the header fields
+	// "content-body": used as the content body of the assertion
+	Statement []byte
+
+	// revision of the new assertion
+	Revision int
+}
+
+func parseStatement(req *SignRequest, dest interface{}) error {
+	switch req.StatementMediaType {
+	case YAMLInput:
+		err := yaml.Unmarshal(req.Statement, dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as YAML: %v", err)
+		}
+	case JSONInput:
+		dec := json.NewDecoder(bytes.NewBuffer(req.Statement))
+		// we want control over supporting only integers
+		dec.UseNumber()
+		err := dec.Decode(dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as JSON: %v", err)
+		}
+	default:
+		return fmt.Errorf("unsupported media type for assertion input: %q", req.StatementMediaType)
+	}
+	return nil
+}
+
+type nestedStatement struct {
+	Headers     map[string]interface{} `yaml:"headers" json:"headers"`
+	ContentBody string                 `yaml:"content-body" json:"content-body"`
+}
+
+// Sign produces the text of a signed assertion as specified by req.
+func Sign(req *SignRequest, keypairMgr asserts.KeypairManager) ([]byte, error) {
+	typ := asserts.Type(req.AssertionType)
+	if typ == nil {
+		return nil, fmt.Errorf("invalid assertion type: %q", req.AssertionType)
+	}
+	if req.Revision < 0 {
+		return nil, fmt.Errorf("assertion revision cannot be negative")
+	}
+	if req.AccountKey == nil && req.KeyID == "" {
+		return nil, fmt.Errorf("both account-key and key id were not specified")
+	}
+
+	var nestedStatement nestedStatement
+	err := parseStatement(req, &nestedStatement)
+	if err != nil {
+		return nil, err
+	}
+	if nestedStatement.Headers == nil {
+		// flat headers, reparse
+		err := parseStatement(req, &nestedStatement.Headers)
+		if err != nil {
+			return nil, err
```

Comment:
So ... the result of this is that it will either fail to parse (and the error will say if it was trying json or yaml) or it will parse but there will still be no headers.

No headers means the signing will fail with something like "type is mandatory" but it might be nicer to tell the user sooner. Perhaps test this IRL and see if the current error reporting is good enough.